cabalscan/src/CabalScan/Options.hs

```diff
@@ -5,32 +5,47 @@ module CabalScan.Options
   , parseCommandLine
   ) where
 
-import Options.Applicative
+import Data.List (intercalate)
 import Path (Abs, File, Path, parseAbsFile)
+import System.Environment
+import System.Exit
+import System.IO (hPutStrLn, stderr)
+
+type AbsoluteFilepath = Path Abs File
 
 data Options = Options
-  { cabalFiles :: [Path Abs File]
+  { cabalFiles :: [AbsoluteFilepath]
   }
 
+data InvocationError = MissingFiles | WrongFilePath String | PrintHelp
+
 parseCommandLine :: IO Options
-parseCommandLine = execParser $
-    info
-      (parser <**> helper)
-      (fullDesc
-         <> progDesc
-              (mconcat
-                 [ "Prints in stdout information extracted from Cabal files"
-                 , " in JSON format."
-                 ]
-              )
-         <> header "cabalscan - extract build information from Cabal files"
-      )
-  where
-    parser = Options
-      <$> some (argument someFile (metavar "CABAL_FILES..."))
+parseCommandLine = do
+  args <- getArgs
+  case parseArgs args [] of
+    Right xs -> return (Options {cabalFiles = xs})
+    Left err -> printMsgAndExit err
 
-    someFile :: ReadM (Path Abs File)
-    someFile = eitherReader $ \s ->
-      case parseAbsFile s of
-        Just f -> Right f
-        Nothing -> Left $ "couldn't parse absolute file path: " ++ s
+parseArgs :: [String] -> [AbsoluteFilepath] -> Either InvocationError [AbsoluteFilepath]
+parseArgs [] [] = Left MissingFiles
+parseArgs [] xs = Right xs
+parseArgs (x:xs) fs | x `elem` ["-h", "--help"] = Left PrintHelp
+                    | otherwise = case parseAbsFile x of
+                                    Just f -> parseArgs xs (f:fs)
+                                    _      -> Left $ WrongFilePath x
+
+printMsgAndExit :: InvocationError -> IO a
+printMsgAndExit = \case
+  MissingFiles -> putErrLn [missing, usage] *> exitFailure
+  WrongFilePath f -> putErrLn [wrongpath f, usage] *> exitFailure
+  PrintHelp -> putErrLn [info, usage, options] *> exitSuccess
```

Comment:
Is the help message supposed to go to stderr?

Reply:
You're right. I've spotted that and it has been fixed in one of the recent commits.